### PR TITLE
Vendor TypeScript and ts-node

### PR DESCRIPTION
**Pull request by julienp**
# Description

Fixes https://github.com/pulumi/pulumi/issues/15733

Historically these packages were direct dependencies of
`@pulumi/pulumi`. To decouple the node SDK from the precise version of
TypeScript, the packages are now declared as optional peer pependencies
of `@pulumi/pulumi` and customers can pick the versions they want.

The reason we mark the peer dependencies as *optional* is to prevent
package managers from automatically installing them. This avoids the
situation where the package manger would install a more recent version
of TypeScript without the user explictly opting in. Newer versions have
stricter type checks, and can thus stop existing programs from running
successfully.

When the peer dependencies are not present, we load the vendored
versions of the modules.

## Checklist

- [ ] I have run `make tidy` to update any new dependencies
- [ ] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: sdk/nodejs/vendor/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+To vendor ts-node:
+
+```bash
+cd sdks/nodejs/vendor
+curl -L -o ts-node-7.0.1.tgz https://registry.npmjs.org/ts-node/-/ts-node-7.0.1.tgz
+tar xvf ts-node-7.0.1.tgz
+cd package
+npm install --omit=dev --no-package-lock --no-bin-links --ignore-scripts
+cd ..
+rsync -r --exclude="*.map" --exclude="*.d.ts" --exclude="*/source-map/dist/*" package/ ts-node@7.0.1/
```
Comment (julienp) [inline sdk/nodejs/vendor/README.md:34]: This vendors the node_modules for ts-node's own dependencies, maybe we can reduce the size a little by creating a bundled version of ts-node using esbuild or rollup.

File: sdk/nodejs/package.json [hunk 2]
```diff
@@ -61,7 +59,21 @@
         "mocha": "^9.0.0",
         "nyc": "^15.1.0",
         "rome": "^12.1.0",
-        "sinon": "^14.0.0"
+        "sinon": "^14.0.0",
+        "ts-node": "^7.0.1",
+        "typescript": "~3.8.3"
```
Comment (julienp) [inline sdk/nodejs/package.json:64]: Make ts-node and typescript *devDependencies* so they are available while working on this repo.

File: sdk/nodejs/tests/runtime/install-package-tests.ts [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            },
+            {
+                typescript: "~3.8.3",
+                "ts-node": "^7.0.1",
+            },
+            {
+                typescript: "^4.0.0",
+            },
+            {
+                typescript: "^5.0.0",
```
Comment (julienp) [inline sdk/nodejs/tests/runtime/install-package-tests.ts:141]: TODO: try a couple more ts-node versions

TODO: @julienp check ts-node performance on later versions https://github.com/pulumi/pulumi/issues/3671#issuecomment-592810579

File: sdk/nodejs/tests/runtime/install-package-tests.ts [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+    await fs.writeFile(path.join(dir, "package.json"), JSON.stringify(packageJSON, undefined, 4));
+}
+
+async function writeTSConfig(dir: string) {
+    const tsconfigJSON = {
+        compilerOptions: {
+            strict: true,
+            target: "es2016",
+            module: "commonjs",
```
Comment (julienp) [inline sdk/nodejs/tests/runtime/install-package-tests.ts:47]: Might be interesting to have an option here in the future for `node16` here to emit ESM

File: sdk/nodejs/tsutils.ts [hunk 5]
```diff
@@ -12,22 +12,56 @@
...
+    let tsnodeRequire = "ts-node";
+
+    try {
+        // Try loading typescript from node_modules
+        const ts: typeof typescript = require("typescript");
+        const tsPath = require.resolve("typescript");
+        log.debug(`Found typescript version ${ts.version} at ${tsPath}`);
+    } catch (error) {
+        // Fallback to the vendored version
+        typescriptRequire = path.join(__dirname, "vendor", "typescript@3.8.3", "typescript.js");
```
Comment (justinvp) [inline sdk/nodejs/tsutils.ts:51]: Woukd it be worth having a `log.debug` for when we're falling back to the vendored version (here and below)?

File: sdk/nodejs/runtime/closure/utils.ts [hunk 6]
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// The  typescript import is used for type-checking only. Do not reference it in the emitted code.
```
Comment (justinvp) [inline sdk/nodejs/runtime/closure/utils.ts:15]: Nit:

```suggestion
// The typescript import is used for type-checking only. Do not reference it in the emitted code.
```

File: sdk/nodejs/runtime/closure/rewriteSuper.ts [hunk 7]
```diff
@@ -12,42 +12,44 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// The  typescript import is used for type-checking only. Do not reference it in the emitted code.
```
Comment (justinvp) [inline sdk/nodejs/runtime/closure/rewriteSuper.ts:15]: Nit:

```suggestion
// The typescript import is used for type-checking only. Do not reference it in the emitted code.
```

File: sdk/nodejs/cmd/run-policy-pack/run.ts [hunk 8]
```diff
@@ -12,11 +12,12 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// The  tsnode import is used for type-checking only. Do not reference it in the emitted code.
```
Comment (justinvp) [inline sdk/nodejs/cmd/run-policy-pack/run.ts:15]: Nit:

```suggestion
// The tsnode import is used for type-checking only. Do not reference it in the emitted code.
```

File: sdk/nodejs/cmd/run/run.ts [hunk 9]
```diff
@@ -12,13 +12,14 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// The  tsnode import is used for type-checking only. Do not reference it in the emitted code.
```
Comment (justinvp) [inline sdk/nodejs/cmd/run/run.ts:15]: Nit:

```suggestion
// The tsnode import is used for type-checking only. Do not reference it in the emitted code.
```

File: sdk/nodejs/runtime/closure/parseFunction.ts [hunk 10]
```diff
@@ -12,10 +12,14 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-import * as ts from "typescript";
+// The  typescript import is used for type-checking only. Do not reference it in the emitted code.
```
Comment (justinvp) [inline sdk/nodejs/runtime/closure/parseFunction.ts:15]: Nit:

```suggestion
// The typescript import is used for type-checking only. Do not reference it in the emitted code.
```

**Review by justinvp — APPROVED** [review]
Some minor things, but otherwise LGTM

Do we need to make any other changes elsewhere before merging this? Like updating templates? Should we give the providers team a heads-up that they can't rely on `@pulumi/pulumi` for pulling-in `tsc` to compile the provider SDK (I believe many (most?) providers are already specifying their own TS dependency).

File: sdk/nodejs/Makefile [hunk 11]
```diff
@@ -35,6 +35,7 @@ build_package:: ensure
 	mkdir -p bin/tests/automation/data/
 	cp -R tests/automation/data/. bin/tests/automation/data/
 	cp README.md ../../LICENSE ./dist/* bin/
+	cp -R ./vendor/ ./bin/vendor/
```
Comment (Frassle) [inline sdk/nodejs/Makefile:38]: `Pack` in the language host probably needs this updating as well

**Comment by julienp** [comment]
> Some minor things, but otherwise LGTM
> 
> Do we need to make any other changes elsewhere before merging this? Like updating templates? Should we give the providers team a heads-up that they can't rely on `@pulumi/pulumi` for pulling-in `tsc` to compile the provider SDK (I believe many (most?) providers are already specifying their own TS dependency).

Templates should be updated after this is released. It's only after we release this that it's safe to use different TypeScript versions.

**Pull request merged**